midend/interpreter.cpp

```diff
@@ -858,17 +858,19 @@ void ExpressionEvaluator::postorder(const IR::Operation_Unary* expression) {
         set(expression, result);
         return;
     }
+    if (auto cast = expression->to<IR::Cast>()) {
```

Comment:
this is not the correct fix, as written for casts constant folding will never happen.
The correct fix is to allow for result below after `auto result = clone->apply(cf)` to return a different type.

Reply:
The problem is that entering constant folding will throw this error. The type is missing despite type inference being applied.

Reply:
this means that there is a second bug, but this is still not the right fix.

Reply:
But really ExpressionEvaluator is an Inspector, there is no reason for the program to have nodes without a type in the tpyeMap.

Reply:
Yes, the bug with the missing type seems unrelated. However, moving cast out like this can probably solve some of the "xzy is not a bool/bit" errors without extra code. Unless `constantFolding` is required here. I do not exactly know the intention of this interpreter code. 

https://github.com/p4lang/p4c/blob/main/frontends/common/constantFolding.cpp#L815

This is the cast that failed, I am not sure how it can happen. Maybe the cast node is generated and not part of the type map?

Reply:
The crash happens here:
```
auto result = clone->apply(cf);
```

and that is because `clone` does not have a type in the typemap, since it was just created. The solution for this is in the previous branch of the `if` statement, which does this:

```
auto type = typeMap->getType(getOriginal(), true);
typeMap->setType(clone, type);  // needed by the constant folding
```

That's why the type isn't there.

But there is another bug; this code assumes that the output type is the same as the input type, which is not true for a cast. I will fix this in a separate PR and we can close this one.